pageserver/src/walreceiver.rs

```diff
@@ -117,38 +120,31 @@ impl<E: Clone> TaskHandle<E> {
     {
         let (cancellation, cancellation_receiver) = watch::channel(());
         let (events_sender, events_receiver) = watch::channel(TaskStateUpdate::Started);
-
-        let join_handle = WALRECEIVER_RUNTIME.spawn(async move {
-            events_sender.send(TaskStateUpdate::Started).ok();
-            task(events_sender, cancellation_receiver).await
+        let events_sender = Arc::new(events_sender);
+        let join_handle = WALRECEIVER_RUNTIME.spawn({
+            let events_sender = Arc::clone(&events_sender);
+            async move {
+                events_sender.send(TaskStateUpdate::Started).ok();
+                let res = task(events_sender, cancellation_receiver).await;
+                res
+            }
         });
 
         TaskHandle {
             join_handle: Some(join_handle),
             events_receiver,
+            _events_sender: events_sender,
             cancellation,
         }
     }
 
     async fn next_task_event(&mut self) -> TaskEvent<E> {
         match self.events_receiver.changed().await {
             Ok(()) => TaskEvent::Update((self.events_receiver.borrow()).clone()),
-            Err(_task_channel_part_dropped) => {
-                TaskEvent::End(match self.join_handle.take() {
-                    Some(jh) => {
-                        if !jh.is_finished() {
-                            warn!("sender is dropped while join handle is still alive");
-                        }
-
-                        jh.await
-                            .map_err(|e| anyhow::anyhow!("Failed to join task: {e}"))
-                            .and_then(|x| x)
-                    }
-                    None => {
-                        // Another option is to have an enum, join handle or result and give away the reference to it
-                        Err(anyhow::anyhow!("Task was joined more than once"))
-                    }
-                })
+            Err(_events_sender_dropped) => {
+                unreachable!(
+                    "we are keeping the sender alive inside self, it can't be dropped here"
+                );
```

Comment:
Before this, if the task finishes without sending an event to `events_receiver`, `next_task_event` would return. With these changes, it will hang forever. That doesn't sound right. Am I missing something?

Reply:
True. Will post a follow-up.